src/log.c

```diff
@@ -47,32 +43,64 @@ char* log_message_timestamp(char* dest, size_t maxlen) {
     return dest;
 }
 
-void log_message_internal(const char* tag, log_level_t level, const char* message, va_list args) {
+void log_message_internal(const char* tag, log_level_t level, const char* message, va_list args, FILE* out) {
     char t_str[LOG_MESSAGE_TIMESTAMP_MAX_LENGTH] = {0};
 
-    if (level > _log_level) {
-        return;
-    }
-
-    fprintf(LOG_MESSAGE_OUTPUT,
+    fprintf(out,
             "[%s][%-11s][%s] ",
             log_message_timestamp(t_str, LOG_MESSAGE_TIMESTAMP_MAX_LENGTH),
             log_level_to_string(level),
             tag);
-    vfprintf(LOG_MESSAGE_OUTPUT, message, args);
-    fprintf(LOG_MESSAGE_OUTPUT, "\n");
-    fflush(LOG_MESSAGE_OUTPUT);
+    vfprintf(out, message, args);
+    fprintf(out, "\n");
+    fflush(out);
 }
 
-void log_set_log_level(log_level_t level) {
-    _log_level = level;
-}
-
-void log_message(const char* tag, log_level_t level, const char* message, ...) {
+void log_message(log_producer_t* tag, log_level_t level, const char* message, ...) {
     va_list args;
     va_start(args, message);
 
-    log_message_internal(tag, level, message, args);
+    for(int i =0; i<= log_service->size-1; ++i){
+        log_sink_t* sink = log_service->sinks[i];
+        if (level > sink->min_level) {
+            continue;
+        }
+        log_message_internal(tag->tag, level, message, args,sink->out);
+    }
 
     va_end(args);
 }
+
+void __attribute__((destructor)) deinit_log_service() {
+    free(log_service->sinks);
+    free(log_service);
+}
+
+void __attribute__((constructor)) init_log_service() {
+    log_service = (log_service_t*)malloc(sizeof(log_service));
+    //init console out
+    log_sink_t* console = init_log_sink(stdout, LOG_LEVEL_INFO);
+    log_service->sinks = malloc(sizeof(log_sink_t));
+    *log_service->sinks = console;
+    log_service->size = 1;
+}
+
+void register_sink(log_sink_t *sink) {
```

Comment:
To match the current naming convention used in cachegrand, can you use a pattern like
log_sink_register

(or for below log_sink_init, log_producer_init, etc.)

Sorry for being annoying but that's the "public" interface of the APIs and it's better to keep it consistent so it's easier to match the expectation of who reads the code (different story for the internal constructors/destructors)